src/ucs/sys/event_set.c

```diff
@@ -0,0 +1,181 @@
+/**
+ * Copyright (C) Hiroyuki Sato. 2019. ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#include "event_set.h"
+
+#include <ucs/debug/memtrack.h>
+#include <ucs/debug/log.h>
+#include <ucs/debug/assert.h>
+
+#include <sys/epoll.h>
+#include <string.h>
+#include <errno.h>
+#include <unistd.h>
+
+#ifdef HAVE_CONFIG_H
+#  include "config.h"
+#endif
+
+#define UCS_EVENT_EPOLL_MAX_EVENTS      16
+#define UCS_EVENT_SET_EPOLL_SIZE 2
+
+struct ucs_event_set_type_table {
+    int event_type;
+    ucs_event_set_type_t os_event;
+};
+
+/* int ucs_event_set_events_map(int events) */
+static int ucs_event_set_events_map(int events)
+{
+    int ret = 0;
+    int i;
+    int matched = 0;
+    struct ucs_event_set_type_table event_table[UCS_EVENT_SET_EPOLL_SIZE] = {
+      { UCS_EVENT_SET_EVREAD,  EPOLLIN  },
+      { UCS_EVENT_SET_EVWRITE, EPOLLOUT },
+    };
+
+    for (i = 0; i < UCS_EVENT_SET_EPOLL_SIZE; i++) {
+       if (events & event_table[i].event_type) {
+         ret |= event_table[i].os_event;
+         matched = 1;
+       }
+    }
+
+    if (!matched)
+        return -1;
+    return ret;
+}
+
+ucs_status_t ucs_event_set_create(ucs_sys_event_set_t **event_set_p)
+{
+    ucs_sys_event_set_t *event_set;
+    ucs_status_t status;
+
+    event_set = ucs_malloc(sizeof(ucs_sys_event_set_t), "ucs_sys_event_set");
+    if (event_set == NULL) {
+        status = UCS_ERR_NO_MEMORY;
+        goto out_create;
+    }
+
+    /* Create epoll set the thread will wait on */
+    event_set->epfd = epoll_create(1);
+    if (event_set->epfd < 0) {
+        ucs_error("epoll_create() failed: %m");
+        status = UCS_ERR_IO_ERROR;
+        goto err_free;
+    }
+
+    status = UCS_OK;
+    *event_set_p = event_set;
+    goto out_create;
+
+err_free:
+    ucs_free(event_set);
+out_create:
+    return status;
+}
+
+ucs_status_t ucs_event_set_add(ucs_sys_event_set_t *event_set, int event_fd,
+                               ucs_event_set_type_t events, void *udata)
+{
+    struct epoll_event event;
+    int ret;
+    int os_events;
+
+    os_events = ucs_event_set_events_map(events);
+    if ( os_events == -1 ) {
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    memset(&event, 0, sizeof(event));
+    event.events   = os_events;
+    event.data.fd  = event_fd;
+    event.data.ptr = udata;
+
+    ret = epoll_ctl(event_set->epfd, EPOLL_CTL_ADD, event_fd, &event);
+    if (ret < 0) {
+        ucs_error("epoll_ctl(epfd=%d, ADD, fd=%d) failed: %m", event_set->epfd,
+                  event_fd);
+        return UCS_ERR_IO_ERROR;
+    }
+
+    return UCS_OK;
+}
+
+ucs_status_t ucs_event_set_mod(ucs_sys_event_set_t *event_set, int event_fd,
+                               ucs_event_set_type_t events, void *udata)
+{
+    struct epoll_event event;
+    int ret;
+    int os_events;
+
+    os_events = ucs_event_set_events_map(events);
+    if ( os_events == -1 ) {
```

Comment:
redundant spaces
also why if nothing matched is an error? maybe the user did pass flags==0?